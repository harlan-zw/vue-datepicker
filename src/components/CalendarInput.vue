<script setup lang="ts">
import type { PropType } from "vue";
import type { Placeholder } from "~/types";

import BaseIcon from "./BaseIcon.vue";

const emit = defineEmits(["open-calendar"]);
defineProps({
  checkIn: {
    type: Date,
    default: null,
  },
  checkOut: {
    type: Date,
    default: null,
  },
  dayFormat: {
    type: Function,
    required: true,
  },
  placeholder: {
    type: Object as PropType<Placeholder>,
    required: true,
  },
});

const openCalendar = () => {
  emit("open-calendar");
};
</script>

<template>
  <div
    data-testid="calendar_input"
    class="calendar_input"
    @click="openCalendar"
  >
    <base-icon
      name="calendar"
      :color="[
        'calendar_input-calendar',
        {
          'calendar_input-calendar--hasnt-checkIn': !checkIn,
          'calendar_input-calendar--checkIn': checkIn,
        },
      ]"
    />

    <p class="calendar_input-text">
      <span
        data-testid="checkIn"
        :class="[
          {
            'calendar_input-text--hasnt-checkIn': !checkIn,
            'calendar_input-text--checkIn': checkIn,
          },
        ]"
      >
        <template v-if="checkIn">
          {{ dayFormat(checkIn) }}
        </template>
        <template v-else>{{ placeholder.checkIn }}</template>
      </span>

      <base-icon
        name="arrowNarrowRight"
        :color="[
          'calendar_input-arrowRight',
          {
            'calendar_input-arrowRight--hasnt-checkIn': !checkIn,
            'calendar_input-arrowRight--checkIn': checkIn,
          },
        ]"
      />

      <span
        data-testid="checkOut"
        :class="[
          {
            'calendar_input-text--hasnt-checkIn': !checkIn,
            'calendar_input-text--checkIn': checkIn,
          },
        ]"
      >
        <template v-if="checkOut">
          {{ dayFormat(checkOut) }}
        </template>
        <template v-else>{{ placeholder.checkOut }}</template>
      </span>
    </p>
  </div>
</template>

<style>
.calendar_input {
  @apply flex items-center h-[50px] cursor-pointer px-4 border border-gray-200;
}
.calendar_input-calendar {
  @apply mr-4;
}
.calendar_input-calendar--hasnt-checkIn {
  @apply text-gray-300;
}
.calendar_input-calendar--checkIn {
  @apply text-gray-700;
}

.calendar_input-text {
  @apply flex items-center m-0;
}
.calendar_input-text--hasnt-checkIn {
  @apply text-gray-300;
}
.calendar_input-text--checkIn {
  @apply text-gray-700;
}

.calendar_input-arrowRight {
  @apply mx-4;
}
.calendar_input-arrowRight--hasnt-checkIn {
  @apply text-gray-300;
}
.calendar_input-arrowRight--checkIn {
  @apply text-gray-700;
}
</style>
