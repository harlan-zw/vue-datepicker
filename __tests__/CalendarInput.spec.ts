import { afterEach, beforeEach, describe, expect, it } from "vitest";

import { format } from "fecha";

import { mount } from "@vue/test-utils";
import CalendarInput from "~/components/CalendarInput.vue";

let wrapper: any;

const dayFormat = (date: Date) => {
  return format(date, "YYYY-MM-DD");
};
const checkIn = new Date();
const checkOut = new Date();

beforeEach(() => {
  wrapper = mount(CalendarInput, {
    props: {
      checkIn,
      checkOut,
      dayFormat,
      placeholder: { checkIn: "Start Date", checkOut: "End date" },
    },
    global: {
      stubs: ["base-icon"],
    },
  });
});

afterEach(() => {
  wrapper.unmount();
});

describe("CalendarInput", () => {
  it("is a Vue instance", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it("should emits open-calendar", async () => {
    const calendarInput = wrapper.get('[data-testid="calendar_input"]');
    await calendarInput.trigger("click");

    expect(wrapper.emitted("open-calendar")).toEqual([[]]);
  });

  it("should render checkIn date", () => {
    const checkInDate = wrapper.get('[data-testid="checkIn"]');

    expect(checkInDate.text()).toBe(dayFormat(checkIn));
  });

  it("should render checkOut date", () => {
    const checkOutDate = wrapper.get('[data-testid="checkOut"]');

    expect(checkOutDate.text()).toBe(dayFormat(checkOut));
  });

  it("should render checkIn placeholder", async () => {
    await wrapper.setProps({ checkIn: null, checkOut: null });
    const checkInDate = wrapper.get('[data-testid="checkIn"]');

    expect(checkInDate.text()).toBe("Start Date");
  });

  it("should render checkOut placeholder", async () => {
    await wrapper.setProps({ checkIn: null, checkOut: null });
    const checkOutDate = wrapper.get('[data-testid="checkOut"]');

    expect(checkOutDate.text()).toBe("End date");
  });
});
