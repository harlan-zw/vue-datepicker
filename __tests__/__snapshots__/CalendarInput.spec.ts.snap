// Vitest Snapshot v1

exports[`CalendarInput > renders CalendarDays properly 1`] = `
"<div data-testid=\\"calendar_input\\" class=\\"calendar_input\\">
  <base-icon-stub name=\\"calendar\\" color=\\"calendar_input-calendar,[object Object]\\" size=\\"xs\\"></base-icon-stub>
  <p class=\\"calendar_input-text\\"><span class=\\"calendar_input-text--checkIn\\"></span>
    <base-icon-stub name=\\"arrowNarrowRight\\" color=\\"calendar_input-arrowRight,[object Object]\\" size=\\"xs\\"></base-icon-stub><span class=\\"calendar_input-text--checkIn\\"></span>
  </p>
</div>"
`;
